File: Python/valid-parentheses.py
# https://leetcode.com/problems/valid-parentheses/
class Solution:
    def isValid(self, s: str) -> bool:
        things = {
            '{':'}',
            '(':')',
            '[':']'
        }

        std = []

        for index,value in enumerate(s):
            if value in things:
                std.append(value)
            elif value in things.values():
                if len(std) > 0:
                    chr = std.pop()
                else:
                    return False
                    
                if things[chr] != value:
                    return False
                else:
                    continue
        if len(std) == 0:
            return True
        else:
            return False


# Optimized Version:
# class Solution:
#     def isValid(self, s: str) -> bool:
#         things = {
#             '{':'}',
#             '(':')',
#             '[':']'
#         }

#         std = []

#         for index,value in enumerate(s):
#             if value in things:
#                 std.append(value)
#             elif len(std) == 0 or things[std.pop()] != value:
#                 return False
#         return len(std) == 0

File: C++/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        map<char,char> brackts={ 
            {'}','{'},{']','['},{')','('}
        };
        stack<char> braces;
        for(int i=0;i<s.size();i++){
            if(s[i]=='(' || s[i]=='[' || s[i]=='{')
                braces.push(s[i]);
            else if(s[i]=='}' ||s[i]==']' ||s[i]==')'){
                if(braces.empty() || (braces.top()!=brackts[s[i]]))
                    return false;
                else
                    braces.pop();
            }
            else
                return false;
        }
        if(braces.empty())
            return true;
        return false;
    }
};